From: Felix Krull <f_krull@gmx.de>
Date: Fri, 27 Apr 2018 08:24:07 -0700
Subject: issue8233

# DP: Issue #8233: When run as a script, py_compile.py optionally takes a single
# DP: argument `-`. Backport from the trunk.
---
 Lib/py_compile.py | 32 ++++++++++++++++++++++++++------
 1 file changed, 26 insertions(+), 6 deletions(-)

diff --git a/Lib/py_compile.py b/Lib/py_compile.py
index 1cb41f1..96bc79a 100644
--- a/Lib/py_compile.py
+++ b/Lib/py_compile.py
@@ -149,16 +149,36 @@ def main(args=None):
     not specified) are compiled and the resulting bytecode is cached
     in the normal manner.  This function does not search a directory
     structure to locate source files; it only compiles files named
-    explicitly.
+    explicitly.  If '-' is the only parameter in args, the list of
+    files is taken from standard input.
 
     """
     if args is None:
         args = sys.argv[1:]
-    for filename in args:
-        try:
-            compile(filename, doraise=True)
-        except PyCompileError,err:
-            sys.stderr.write(err.msg)
+    rv = 0
+    if args == ['-']:
+        while True:
+            filename = sys.stdin.readline()
+            if not filename:
+                break
+            filename = filename.rstrip('\n')
+            try:
+                compile(filename, doraise=True)
+            except PyCompileError, error:
+                rv = 1
+                sys.stderr.write("%s\n" % error.msg)
+            except IOError, error:
+                rv = 1
+                sys.stderr.write("%s\n" % error)
+    else:
+        for filename in args:
+            try:
+                compile(filename, doraise=True)
+            except PyCompileError, error:
+                # return value to indicate at least one failure
+                rv = 1
+                sys.stderr.write(error.msg)
+    return rv
 
 if __name__ == "__main__":
     main()
